pkg/controllermanager/controller/plant/plant_control.go

```diff
@@ -0,0 +1,357 @@
+			return err
+		}
+	}
+
+	var (
+		newConditions               = helper.MergeConditions(plant.Status.Conditions, helper.InitCondition(gardencorev1alpha1.PlantAPIServerAvailable), helper.InitCondition(gardencorev1alpha1.PlantEveryNodeReady))
+		conditionAPIServerAvailable = newConditions[0]
+		conditionEveryNodeReady     = newConditions[1]
+	)
+
+	kubeconfigSecret, err := c.secretsLister.Secrets(plant.Namespace).Get(plant.Spec.SecretRef.Name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return c.updateStatusToUnknown(plant, "Referenced Plant secret could not be found.", conditionAPIServerAvailable, conditionEveryNodeReady)
+		}
+		return err
+	}
+
+	secretFinalizers := sets.NewString(kubeconfigSecret.Finalizers...)
+	if !secretFinalizers.Has(FinalizerName) {
+		secretFinalizers.Insert(FinalizerName)
+	}
+	kubeconfigSecret.Finalizers = secretFinalizers.UnsortedList()
+	if err := c.k8sGardenClient.Client().Update(ctx, kubeconfigSecret); err != nil {
+		return err
+	}
+
+	kubeconfig, ok := kubeconfigSecret.Data["kubeconfig"]
+	if !ok {
+		message := "Plant secret needs to contain a kubeconfig key."
+		return c.updateStatusToUnknown(plant, message, conditionAPIServerAvailable, conditionEveryNodeReady)
+	}
+
+	plantClusterClient, discoveryClient, err := c.initializePlantClients(plant, key, kubeconfig)
+	if err != nil {
+		message := fmt.Sprintf("Could not initialize Plant clients: %+v", err)
+		return c.updateStatusToUnknown(plant, message, conditionAPIServerAvailable, conditionEveryNodeReady)
+	}
+
+	healthChecker := c.initializeHealthChecker(plantClusterClient, discoveryClient)
+
+	// Trigger health check
+	conditionAPIServerAvailable, conditionEveryNodeReady = c.healthChecks(ctx, healthChecker, logger, conditionAPIServerAvailable, conditionEveryNodeReady)
+
+	cloudInfo, err := FetchCloudInfo(ctx, plantClusterClient, discoveryClient, logger)
+	if err != nil {
+		return err
+	}
+
+	return c.updateStatus(plant, cloudInfo, conditionAPIServerAvailable, conditionEveryNodeReady)
+}
+
+func (c *defaultPlantControl) updateStatusToUnknown(plant *gardencorev1alpha1.Plant, message string, conditionAPIServerAvailable, conditionEveryNodeReady gardencorev1alpha1.Condition) error {
+	conditionAPIServerAvailable = helper.UpdatedCondition(conditionAPIServerAvailable, gardencorev1alpha1.ConditionFalse, "APIServerDown", message)
+	conditionEveryNodeReady = helper.UpdatedCondition(conditionEveryNodeReady, gardencorev1alpha1.ConditionFalse, "Nodes not reachable", message)
+	return c.updateStatus(plant, &StatusCloudInfo{}, conditionAPIServerAvailable, conditionEveryNodeReady)
+}
+
+func (c *defaultPlantControl) updateStatus(plant *gardencorev1alpha1.Plant, cloudInfo *StatusCloudInfo, conditions ...gardencorev1alpha1.Condition) error {
+
+	cur, err := c.k8sGardenClient.GardenCore().CoreV1alpha1().Plants(plant.ObjectMeta.Namespace).Get(plant.ObjectMeta.Name, metav1.GetOptions{})
```

Comment:
What's the reason to read the `Plant` again from the system?
Also, if use the controller-runtime's client in this function (and remove empty line at beginning of function)